### ISSUE TITLE
[i18n] implement i18n strategy at component level, modularly

### ISSUE THREAD

**Issue by btopro** [elmsln/issues#141]
After trying 4 different approaches they seem to suffer from the following:
- poor DX
- if great DX, than assumes a singular application / 1 file w/ all translations
- if great DX (lit core one) then assumes using LitElement / lit-html at all times (we use Vanilla / others at times)

While thinking on the shortcomings of what exists while trying to rethink how we do translation, I thought about this possible workflow. What if we had a singleton  that was JUST in charge of translation. Following a paradigm @nikkimk came up with in responsive-utility; what if we try the following for language as she did for responsive / breakpoint communication:
- Things emit event if they care about translation
- Text is in a variable / object internal to element
- Hook / reference to the path for WHERE the translation file is per language, also emits what languages it supports.
- On language change as monitored by the singleton, it sends back to all references the new object values when lang file processed by the singleton.
- We ONLY translate if we have things to translate internal to the element and a file w. Translations object
- Also bubble up a callback post variables changing otherwise assumes lit if requestUpdate is defined 
- Update lang call MUST be async
- 0 dependency convention or is it a mix in? Both maybe?

Using this type of pattern we could potentially load multiple `.json` files relative to the language being changed. Things that an element wanting into this would have to follow (which could then be convention as well as mix in)
- `this.elText` or some `Object` which supplies the strings so that in LitElement we could write `${this.elText.title}` and get the same immediate updating if `this.elText.title` changes but that we could also use this in non-lit projects to similar affect
- event bubble up needs to supply a callback for processing when the text strings change so that the singleton can do the following:
  1. Know who to update
  2. On lang change, see which elements supplied a related whatever.json for translations and import the file if not in local cache
  3. Update an associated `this.elText` or comparable variable to rewrite the matches in there
  4. Run a callback function of some kind (like `this.requestUpdate` in LitElement to force repaint or `this.render()` in most of our vanilla stuff)
- With who to update, what text to update, where to load that text and what to call after to reflect the change, I think this *could* work in the lazy assembled way we are hoping for
- Need to ensure that multiple things registering translation (like a simple-button) would ONLY register 1x translation .json yet beam back to update all of them successfully
- On language change, need to run in async as not to cripple things and also probably want to load default english so that internal state management stuff sets up the dom and then update language right after as nessecary

This gets us away from dependency trees of other things that really just provide this in the end. We should also try to support `en` and `en-US` and `en-UK` style notations. If an element supplies a `en` translation, the page being lang `en-US` must still load that one (default in this case so loads nothing). If the element supplies a `en-UK` variant and the page is `en-UK` then it loads in default english (as per all elements) and then does the processing to change JUST the elements that have a specific UK variation.

- [x] If a variation is not set like `ja` and the element supplies a `ja` or `ja-SP` (any variation that is) then it needs to use that variation. 
- [x] If it does NOT supply the translation it needs to keep things the same and not fail
- [x] Need a simple, documented way of generating these translation files and in a consistent `locales` directory (or whatever) so our tooling internally doesn't choke on it either

**Comment by btopro** [comment]
looking at https://github.com/RolandDreger/web-components/blob/master/note-list/src/note-list.js it brought me back to how much I liked the idea of modular, small json blobs of translations. I do wonder what the performance impact of this is but it would be very easy to maintain / contribute updates to.

Reading through https://github.com/andreasbm/lit-translate it looks like I need to implement it in a few things and see how it starts to play out. Methodology I want to leverage for translations:
- [ ] translations live per element (everything is still pieces, unbundled; need to ensure this updates without crippling things in larger context)
- [ ] tooling needs to pick up and find this location accurately
- [ ] whatever we use (lit-translate looks great) it should be an abstraction into our utils project / its own repo. This way we can abstract and point to our own to simplify a later rewrite if it occured
- [ ] translations are fetched as `something/locales/element-name.langCode.json`
- [ ] translations are automatically updated / reactive if language changes. While this serves no practical purpose, it will ensure that our DX is solid as I love the idea of just writing `translate("var.name")` and it translates the property correctly.
- [ ] We might need to differentiate user defined text like a `this.buttonName` being translated vs `label="Cool button"` being hard coded. I think what we're solving will be the hard coded part. Will need to look into the idea of the element implementing `this.buttonName` needing to be in charge of managing translations at its level
- [ ] I hope this isn't crippling but I have a feeling `en-UK` fetching `whatever.json` per element might be pretty aggressive and laggy but we'll see. Catching defaulting to always use what's already stored locally might be a good way around this. starting with HAX UI layer and working forward from there as far as implementation could be a good way to battle test it
- [ ] We'll also want to see if this conflicts in any way w/ services like those built into Chrome that just translate things automatically

**Comment by btopro** [comment]
This is in enough of a state we can ship it, test and refine further :). Very easy to wire up and works for other languages as supported by the elements.
